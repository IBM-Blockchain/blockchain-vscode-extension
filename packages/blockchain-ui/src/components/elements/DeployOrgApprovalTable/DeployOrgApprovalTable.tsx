import React, { FunctionComponent } from 'react';
import { DataTable, InlineNotification, TableContainer, Table, TableHead, TableRow, TableHeader, TableBody, TableCell } from 'carbon-components-react';

interface IProps {
    orgApprovals: any;
    orgMap: any;
    environmentPeers: string[];
}

const DeployOrgApprovalTable: FunctionComponent<IProps> = ({orgApprovals, orgMap, environmentPeers}) => {
    const headerData: { header: string, key: string }[] = [
        {
            header: 'Organization',
            key: 'organization',
        },
        {
            header: 'Approval status',
            key: 'status',
        }
    ];

    let tableElement: JSX.Element = <></>;

    if (orgApprovals) {
        const rowData: { id: string, organization: string, status: string }[] = [];

        const approvalEntries: any[] = Object.entries(orgApprovals);

        if (approvalEntries.length > 0) {
            for (const [org, approved] of approvalEntries) {

                let environmentPeer: string | undefined;
                if (!approved) {
                    // Org should be marked as pending if we're about to deploy on an organization's peers.
                    environmentPeer = orgMap[org].find((_peer: string) => environmentPeers.includes(_peer));
                }

                const entry: { id: string, organization: string, status: string } = {
                    id: org,
                    organization: org,
                    status: approved ? 'Approved' : environmentPeer ? 'Pending (part of this deploy)' : 'Not approved'
                };

                rowData.push(entry);

                tableElement = (
                    <DataTable
                        rows={rowData}
                        headers={headerData}
                        render={({ rows, headers, getHeaderProps }) => (
                            <TableContainer>
                                <Table size='short'>
                                    <TableHead>
                                        <TableRow>
                                            {headers.map((header) => (
                                                <TableHeader {...getHeaderProps({ header })}>
                                                    {header.header}
                                                </TableHeader>
                                            ))}
                                        </TableRow>
                                    </TableHead>
                                    <TableBody>
                                        {rows.map((row) => (
                                            <TableRow id={row.id + '-row'} key={row.id}>
                                                {row.cells.map((cell) => (
                                                    <TableCell key={cell.id}>{cell.value}</TableCell>
                                                ))}
                                            </TableRow>
                                        ))}
                                    </TableBody>
                                </Table>
                            </TableContainer>)}
                    />
                );
            }
        } else {
            tableElement = (
                <InlineNotification
                    hideCloseButton={true}
                    kind='info'
                    lowContrast={true}
                    notificationType='inline'
                    role='alert'
                    statusIconDescription='describes the status icon'
                    subtitle={<p>Commit has already been performed for this definition name and version.</p>}
                    title='Unable to get organization approvals'
                />
            );
        }
    } else {
        tableElement = (<p>Retrieving organization approvals...</p>);
    }

    return tableElement;
};

export default DeployOrgApprovalTable;
