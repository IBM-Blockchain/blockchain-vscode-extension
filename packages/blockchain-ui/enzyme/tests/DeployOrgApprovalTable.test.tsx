import React from 'react';
import renderer from 'react-test-renderer';
import chai from 'chai';
import sinonChai from 'sinon-chai';
import DeployOrgApprovalTable from '../../src/components/elements/DeployOrgApprovalTable/DeployOrgApprovalTable';
import { ShallowWrapper, shallow } from 'enzyme';
chai.should();
chai.use(sinonChai);

// tslint:disable no-unused-expression

describe('DeployOrgApprovalTable component', () => {
    let defaultProps: { orgApprovals: any, orgMap: any, environmentPeers: string[] };

    beforeEach(() => {
        defaultProps = {
            orgApprovals: {
                Org1MSP: true,
                Org2MSP: false
            },
            orgMap: {
                Org1MSP: ['Org1Peer1'],
                Org2MSP: ['Org2Peer1']
            },
            environmentPeers: [
                'Org1Peer1'
            ]
        };
    });

    it('should render the expected snapshot', () => {
        const component: any = renderer
            .create(<DeployOrgApprovalTable {...defaultProps} />)
            .toJSON();
        expect(component).toMatchSnapshot();
    });

    it('should show which orgs have approved the definition', () => {
        const component: ShallowWrapper<any> = shallow(<DeployOrgApprovalTable {...defaultProps}/>);
        component.html().includes(`<tr id="Org1MSP-row"><td>Org1MSP</td><td>Approved</td></tr><tr id="Org2MSP-row"><td>Org2MSP</td><td>Not approved</td></tr>`).should.equal(true);
        component.html().includes(`<p>Commit has already been performed for this definition name and version.</p>`).should.equal(false);
    });

    it('should show which orgs will approve the definition when deployed', () => {
        defaultProps.orgApprovals = {
            Org1MSP: false,
            Org2MSP: false
        };
        const component: ShallowWrapper<any> = shallow(<DeployOrgApprovalTable {...defaultProps}/>);
        component.html().includes(`<tr id="Org1MSP-row"><td>Org1MSP</td><td>Pending (part of this deploy)</td></tr><tr id="Org2MSP-row"><td>Org2MSP</td><td>Not approved</td></tr>`).should.equal(true);
        component.html().includes(`<p>Commit has already been performed for this definition name and version.</p>`).should.equal(false);
    });

    it('should show error if unable to get org approvals', () => {
        defaultProps.orgApprovals = {};
        const component: ShallowWrapper<any> = shallow(<DeployOrgApprovalTable {...defaultProps}/>);
        component.html().includes(`<p>Commit has already been performed for this definition name and version.</p>`).should.equal(true);
        component.html().includes(`<tr><td>Org1MSP</td><td>Approved</td></tr><tr><td>Org2MSP</td><td>Not approved</td></tr>`).should.equal(false);
    });

    it('should not show org approval table if no response from vs code', () => {
        defaultProps.orgApprovals = undefined;
        const component: ShallowWrapper<any> = shallow(<DeployOrgApprovalTable {...defaultProps}/>);
        component.html().includes(`<p>Retrieving organization approvals...</p>`).should.equal(true);
    });
});
