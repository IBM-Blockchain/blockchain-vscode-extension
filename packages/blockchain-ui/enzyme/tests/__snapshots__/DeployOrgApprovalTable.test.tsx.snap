// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeployOrgApprovalTable component should render the expected snapshot 1`] = `
<div
  className="bx--data-table-container"
>
  <table
    className="bx--data-table bx--data-table--short bx--data-table--no-border"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          <span
            className="bx--table-header-label"
          >
            Organization
          </span>
        </th>
        <th
          scope="col"
        >
          <span
            className="bx--table-header-label"
          >
            Approval status
          </span>
        </th>
      </tr>
    </thead>
    <tbody
      aria-live="polite"
    >
      <tr
        id="Org1MSP-row"
      >
        <td>
          Org1MSP
        </td>
        <td>
          Approved
        </td>
      </tr>
      <tr
        id="Org2MSP-row"
      >
        <td>
          Org2MSP
        </td>
        <td>
          Not approved
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
